numba/typing/cffi_utils.py

```diff
@@ -92,42 +95,86 @@ def _type_map():
     return _cached_type_map
 
 
-def map_type(cffi_type):
+def map_type(cffi_type, use_record_dtype=False):
     """
     Map CFFI type to numba type.
+
+    Parameters
+    ----------
+    cffi_type:
+        The CFFI type to be converted.
+    use_record_dtype: bool (default: False)
+        When True, struct types are mapped to numpy record dtype.
+
     """
+    primed_map_type = partial(map_type, use_record_dtype=use_record_dtype)
     kind = getattr(cffi_type, 'kind', '')
     if kind == 'union':
         raise TypeError("No support for CFFI union")
     elif kind == 'function':
         if cffi_type.ellipsis:
             raise TypeError("vararg function is not supported")
-        restype = map_type(cffi_type.result)
-        argtypes = [map_type(arg) for arg in cffi_type.args]
+        restype = primed_map_type(cffi_type.result)
+        argtypes = [primed_map_type(arg) for arg in cffi_type.args]
         return templates.signature(restype, *argtypes)
     elif kind == 'pointer':
         pointee = cffi_type.item
         if pointee.kind == 'void':
             return types.voidptr
         else:
-            return types.CPointer(map_type(pointee))
+            return types.CPointer(primed_map_type(pointee))
     elif kind == 'array':
-        return map_type(cffi_type.item)
+        return primed_map_type(cffi_type.item)
+    elif kind == 'struct' and use_record_dtype:
+        return map_struct_to_record_dtype(cffi_type)
     else:
         result = _type_map().get(cffi_type)
         if result is None:
             raise TypeError(cffi_type)
         return result
 
 
-def make_function_type(cffi_func):
+def map_struct_to_record_dtype(cffi_type):
+    """Convert a cffi type into a numpy record dtype
+    """
+    fields = {
+            'names': [],
+            'formats': [],
+            'offsets': [],
+            'itemsize': ffi.sizeof(cffi_type),
+    }
+    is_aligned = True
+    for k, v in cffi_type.fields:
+        # guard unsupport values
+        if v.bitshift != -1:
+            msg = "fields with bitshift are not supported: {}"
+            raise ValueError(msg.format(k))
+        if v.flags != 0:
+            msg = "fields has unsupport flags: {}"
```

Comment:
Think this needs altering? Any flag is unsupported opposed to there being some flags which are unsupported?
```suggestion
            msg = "field {} has flags, this is not supported"
```